<?xml version="1.0" encoding="UTF-8"?>
<launch>
  <arg name="manager_name" default="nodelet_manager" />
  <arg name="nodelet_mode" default="standalone" /><!-- set to standalone if you want to use as node-->
  <arg name="debug" default="true"/>
  <arg name="robot_name" default="omni4"/>
  <arg name="sim" default="false"/>

  <include file="$(eval find(robot_name) + '/launch/RobotHW.launch' )">
    <arg name="manager_name" value="$(arg manager_name)" />
    <arg name="nodelet_mode" value="$(arg nodelet_mode)" />
    <arg name="robot_name" value="$(arg robot_name)"/>
  </include>
  

  <group ns="$(arg robot_name)">
        <node pkg="nodelet" type="nodelet" name="odom_broadcaster" machine="robot"
    args="$(arg nodelet_mode) odometry_source_plugins/OdomBroadcaster $(arg manager_name)" output="screen">
      <param name="odom_frame" value="$(arg robot_name)/odom" />
      <param name="base_frame"  value="$(arg robot_name)/base_link" />
      <param name="publish_rate"  value="1000" />
      <param name="x0" value="0.55"/>
      <param name="y0" value="9.55"/>
      <remap from="odom/x" to="base/odom/x"/>
      <remap from="odom/y" to="base/odom/y"/>
      <remap from="odom/yaw" to="base/odom/yaw"/>
    </node>

    <node pkg="ds4_driver" type="ds4_driver_node.py" name="ds4_driver" output="screen" machine="robot">
      <rosparam command="load" file="$(find ds4_driver)/config/params.yaml" />
      <param name="device_addr" value="" />
      <param name="use_standard_msgs" value="true" />
    </node>

    <node pkg="map_server" type="map_server" name="map_server" args="$(find robot_control)/resources/red/map.yaml">
      <param name="frame_id" value="$(arg robot_name)/map"/>
    </node>

      <node pkg="amcl" type="amcl" name="amcl" output="screen">
        <rosparam command="load" file="$(find robot_control)/config/amcl.yaml" />
        <param name="base_frame_id" value="$(arg robot_name)/base_link" />
        <param name="odom_frame_id" value="$(arg robot_name)/odom" />
        <param name="global_frame_id" value="$(arg robot_name)/map" />
        <param name="initial_pose_x" value="0.55" />
        <param name="initial_pose_y" value="9.55" />
        <param name="initial_pose_a" value="0" />
      </node>


    <node pkg="move_base" type="move_base" respawn="false" name="move_base" output="screen">
      <rosparam file="$(find robot_control)/config/move_base/dwa_local_planner_params.yaml" command="load" ns="DWAPlannerROS" />
      <rosparam file="$(find robot_control)/config/move_base/costmap_common_params.yaml" command="load" ns="global_costmap" />
      <rosparam file="$(find robot_control)/config/move_base/costmap_common_params.yaml" command="load" ns="local_costmap" />
      <rosparam file="$(find robot_control)/config/move_base/local_costmap_params.yaml" command="load" />
      <rosparam file="$(find robot_control)/config/move_base/global_costmap_params.yaml" command="load" />

      <param name="global_costmap/laser_scan_sensor/sensor_frame" value="$(arg robot_name)/laser_link" />
      <param name="global_costmap/laser_scan_sensor/topic" value="scan" />
      <param name="global_costmap/global_frame" value="$(arg robot_name)/map" />
      <param name="global_costmap/robot_base_frame" value="$(arg robot_name)/base_link" />

      <param name="local_costmap/laser_scan_sensor/sensor_frame" value="$(arg robot_name)/laser_link" />
      <param name="local_costmap/laser_scan_sensor/topic" value="scan" />
      <param name="local_costmap/global_frame" value="$(arg robot_name)/odom" />
      <param name="local_costmap/robot_base_frame" value="$(arg robot_name)/base_link" />

      <!-- <remap from="odom" to="odom_twist"/> -->

      <param name="base_local_planner" value="dwa_local_planner/DWAPlannerROS" />
     </node>
  </group>


  <!-- Debug -->
  <group if="$(arg debug)">
    <include file="$(find robot_control)/launch/debug.launch">
      <arg name="robot_name" value="$(arg robot_name)"/>
    </include>
  </group>
 
</launch>